src/tap/src/tapcell.cpp

```diff
@@ -33,41 +33,1306 @@
+                  const char* tap_nwout3_master,
+                  const char* incnrcap_nwin_master,
+                  const char* incnrcap_nwout_master,
+                  const char* tapcell_master,
+                  int& dist)
+{
+  cutRows(endcap_master, findBlockages(), halo_x, halo_y);
+  vector<vector<odb::dbRow*>> rows = organizeRows();
+  vector<string> cnrcap_masters;
+  string cnrcap_nwin_master_string(cnrcap_nwin_master);
+  string cnrcap_nwout_master_string(cnrcap_nwout_master);
+  cnrcap_masters.push_back(cnrcap_nwin_master_string);
+  cnrcap_masters.push_back(cnrcap_nwout_master_string);
+
+  phy_idx_ = 0;
+  filled_sites_.clear();
+  if (endcap_master != nullptr) {
+    insertEndcaps(rows, endcap_master, cnrcap_masters);
+  }
+  if (add_boundary_cell) {
+    vector<string> tap_nw_masters;
+    tap_nw_masters.push_back(tap_nwintie_master);
+    tap_nw_masters.push_back(tap_nwin2_master);
+    tap_nw_masters.push_back(tap_nwin3_master);
+    tap_nw_masters.push_back(tap_nwouttie_master);
+    tap_nw_masters.push_back(tap_nwout2_master);
+    tap_nw_masters.push_back(tap_nwout3_master);
+
+    vector<string> tap_macro_masters;
+    tap_macro_masters.push_back(incnrcap_nwin_master);
+    tap_macro_masters.push_back(tap_nwin2_master);
+    tap_macro_masters.push_back(tap_nwin3_master);
+    tap_macro_masters.push_back(tap_nwintie_master);
+    tap_macro_masters.push_back(incnrcap_nwout_master);
+    tap_macro_masters.push_back(tap_nwout2_master);
+    tap_macro_masters.push_back(tap_nwout3_master);
+    tap_macro_masters.push_back(tap_nwouttie_master);
+
+    insertAtTopBottom(rows,
+                      tap_nw_masters,
+                      db_->findMaster(cnrcap_nwin_master),
+                      endcap_prefix_);
+    insertAroundMacros(rows,
+                       tap_macro_masters,
+                       db_->findMaster(cnrcap_nwin_master),
+                       endcap_prefix_);
+  }
+  insertTapcells(rows, tapcell_master, dist);
+  filled_sites_.clear();
+}
+
+void Tapcell::cutRows(odb::dbMaster* endcap_master,
+                      vector<odb::dbBox*> blockages,
+                      int& halo_x,
+                      int& halo_y)
+{
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  int rows_count = (block->getRows()).size();
+  int block_count = blockages.size();
+  int min_row_width;
+  int end_width;
+
+  if (endcap_master != nullptr) {
+    end_width = endcap_master->getWidth();
+    min_row_width = 2 * end_width;
+  } else {
+    min_row_width = 0;
+  }
+
+  // Gather rows needing to be cut up front
+  vector<odb::dbRow*> blocked_rows;
+  std::map<string, vector<odb::dbBox*>> row_blockages;
+  string row_name;
+  for (odb::dbBox* blockage : blockages) {
+    for (odb::dbRow* row : block->getRows()) {
+      if (overlaps(blockage, row, halo_x, halo_y)) {
+        row_name = row->getName();
+        if (row_blockages.find(row_name) == row_blockages.end()) {
+          blocked_rows.push_back(row);
```

Comment:
There is no need to use the row name,  just use dbRow* which is unique and faster.  You can also eliminated blocked_rows as the keys of row_blockages are the rows.

Reply:
I adjusted the row name and how the values are inserted into the map as advised in the other comments. However if the keys of row_blockages are used instead of blocked_rows, the boundary_macros test fails as it finds differences in the def file. I believe this is because the keys in the map are arranged in ascending order automatically as opposed to the order the rows are being inserted in the blocked_rows vector.